### PR TITLE
Implement support for PNS (Process Namespace Sharing) executor

### PR DESCRIPTION
**Pull request by jessesuen**
Issue https://github.com/argoproj/argo/issues/970

* Implements PNS (Process Namespace Sharing) executor
* Adds limited support for Kubelet/K8s API artifact collection by mirroring volume mounts to wait sidecar
* Adds validation to detect when output artifacts are not supported by the executor
* Adds ability to customize executor from workflow-controller-configmap (e.g. add environment variables, append command line args such as loglevel)
* Fixes an issue where daemon steps were not getting terminated properly

### CODE REVIEW

File: util/archive/archive_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	w := bufio.NewWriter(f)
+
+	err = TarGzToWriter(data.Name(), w)
+	assert.Nil(t, err)
+	err = os.Remove(data.Name())
+	assert.Nil(t, err)
+
+	err = f.Close()
+	assert.Nil(t, err)
+}
```
Comment (jessesuen) [inline util/archive/archive_test.go]: Add more unit tests for corner cases: 
* symlink in directory
* file which is a symlink

File: util/archive/archive.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		return errors.InternalWrapError(err)
+	}
+	if !sourceFi.Mode().IsRegular() && !sourceFi.IsDir() {
+		return errors.InternalErrorf("%s is not a regular file or directory", sourcePath)
+	}
+	if flush, ok := w.(flusher); ok {
+		defer flush.Flush()
+	}
+	gzw := gzip.NewWriter(w)
+	defer gzw.Close()
```
Comment (jessesuen) [inline util/archive/archive.go]: Fix linting
```
defer func() { _ = gzw.Close() }
```

File: workflow/controller/config.go [hunk 3]
```diff
@@ -22,11 +22,16 @@ import (
...
 
 	// ExecutorImagePullPolicy is the imagePullPolicy of the executor to use when running pods
+	// DEPRECATED: use `executor.imagePullPolicy` in configmap instead
 	ExecutorImagePullPolicy string `json:"executorImagePullPolicy,omitempty"`
 
+	// Executor  executor to use when running pods
+	Executor *apiv1.Container `json:"executor,omitempty"`
+
 	// ExecutorResources specifies the resource requirements that will be used for the executor sidecar
 	ExecutorResources *apiv1.ResourceRequirements `json:"executorResources,omitempty"`
```
Comment (jessesuen) [inline workflow/controller/config.go]: DEPECRATE executor resources in favor of Executor config field

File: workflow/executor/pns/pns.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+	for _, proc := range allProcs {
+		pid := proc.Pid()
+		if pid == 1 || pid == thisPid || proc.PPid() != 0 {
+			// ignore the pause container, our own pid, and non-root processes
+			continue
+		}
+		if p.mainFS == nil {
+			// if we have yet to secure the main container's filesystem, grab all root filesystems
+			if _, ok := p.pidFileHandles[pid]; !ok {
+				rootPath := fmt.Sprintf("/proc/%d/root", pid)
```
Comment (dtaniwaki) [inline workflow/executor/pns/pns.go]: This is not possible in my environment.

Could you try the following manifest and try to browse the file system of the main container from the sidecar?


```
kind: Pod
apiVersion: v1
metadata:
  name: pns
spec:
  containers:
  - name: main
    image: ubuntu:16.04
    command: [sleep, infinity]
    securityContext:
      runAsUser: 1234
  - name: sidecar
    image: ubuntu:16.04
    command: [sleep, infinity]
    securityContext:
      runAsUser: 0
  shareProcessNamespace: true
```

Even if you are root, you can't browse it. Do I have to set up something to allow it?

File: workflow/validate/validate.go [hunk 5]
```diff
@@ -514,6 +531,51 @@ func validateOutputs(scope map[string]interface{}, tmpl *wfv1.Template) error {
 	return nil
 }
 
+// validateBaseImageOutputs detects if the template contains an output from
+func (ctx *wfValidationCtx) validateBaseImageOutputs(tmpl *wfv1.Template) error {
```
Comment (dtaniwaki) [inline workflow/validate/validate.go]: The name of this function sounds a little bit unclear for me. Does it validate whether outputs are valid? Maybe, the meaning of `BaseImage` is ambiguous. What do you think?

File: workflow/validate/validate.go [hunk 6]
```diff
@@ -514,6 +531,51 @@ func validateOutputs(scope map[string]interface{}, tmpl *wfv1.Template) error {
...
+	switch ctx.ContainerRuntimeExecutor {
+	case "", common.ContainerRuntimeExecutorDocker:
+		// docker executor supports all modes of artifact outputs
+	case common.ContainerRuntimeExecutorPNS:
+		// pns supports copying from the base image, but only if there is no volume mount underneath it
+		errMsg := "pns executor does not support outputs from base image layer with volume mounts. must use emptyDir"
```
Comment (dtaniwaki) [inline workflow/validate/validate.go]: Are we actually checking the volume type is emptyDir here?

File: workflow/controller/workflowpod.go [hunk 7]
```diff
@@ -539,6 +544,41 @@ func (woc *wfOperationCtx) addInputArtifactsVolumes(pod *apiv1.Pod, tmpl *wfv1.T
...
+		}
+	}
+	if mainCtrIndex == -1 || waitCtrIndex == -1 {
+		panic("Could not find main or wait container in pod spec")
+	}
+	mainCtr = &pod.Spec.Containers[mainCtrIndex]
+	waitCtr := &pod.Spec.Containers[waitCtrIndex]
+
+	for _, mnt := range mainCtr.VolumeMounts {
+		mnt.MountPath = filepath.Join(common.ExecutorMainFilesystemDir, mnt.MountPath)
```
Comment (dtaniwaki) [inline workflow/controller/workflowpod.go]: `filepath.Join("/a/b", "../../c/d")` becomes `/c/d` while `filepath.Join("/a/b", "/c/d")` becomes `/a/b/c/d`. Can we use volume mounts index mapping to make it more secure?

File: workflow/executor/pns/pns.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+// pollRootProcesses will poll /proc for root pids (pids without parents) in a tight loop, for the
+// purpose of securing an open file handle against /proc/<pid>/root as soon as possible.
+// It opens file handles on all root pids because at this point, we do not yet know which pid is the
+// "main" container.
+// Polling is necessary because it is not possible to use something like fsnotify against procfs.
+func (p *PNSExecutor) pollRootProcesses(timeout time.Duration) {
+	log.Warnf("Polling root processes (%v)", timeout)
+	deadline := time.Now().Add(timeout)
+	for {
+		p.updateCtrIDMap()
```
Comment (alexmt) [inline workflow/executor/pns/pns.go]: should executor panic if `updateCtrIDMap` returns error?

File: docs/workflow-controller-configmap.yaml [hunk 9]
```diff
@@ -71,16 +71,27 @@ data:
     # disable the TLS verification of the kubelet executo (default: false)
```
Comment (alexmt) [inline docs/workflow-controller-configmap.yaml]: can you document possible values for `containerRuntimeExecutor` ?

File: workflow/executor/pns/pns.go (hunk 4 above)
Comment (alexmt) [reply workflow/executor/pns/pns.go]: I think I'm seeing same behavior. [artifact-passing.yaml](https://github.com/argoproj/argo/blob/master/examples/artifact-passing.yaml) fails even if I add 'sleep 10' to generate-artifact step. According to logs executor was able to save artifacts and then failed because could not chroot into main:

```
time="2019-02-13T18:46:08Z" level=info msg="Saving output artifacts"
time="2019-02-13T18:46:08Z" level=info msg="Staging artifact: hello-art"
time="2019-02-13T18:46:08Z" level=info msg="Copying /tmp/hello_world.txt from container base image layer to /argo/outputs/artifacts/hello-art.tgz"
time="2019-02-13T18:46:08Z" level=error msg="executor error: could not chroot into main for artifact collection: container may have exited too quickly
```

File: workflow/artifacts/s3/s3.go [hunk 10]
```diff
@@ -85,11 +88,14 @@ func (s3Driver *S3ArtifactDriver) Save(path string, outputArtifact *wfv1.Artifac
 			}
 			if isDir {
 				if err = s3cli.PutDirectory(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
+					log.Warnf("Failed to put directory: %v", err)
+					return false, nil
+				}
+			} else {
+				if err = s3cli.PutFile(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
```
Comment (maxkramer) [inline workflow/artifacts/s3/s3.go]: As discussed on Slack, would it be worth adding an info log to stat the file size for debug purposes, in the event that we see a discrepancy between the file on disk and the file in the S3 bucket?

File: workflow/controller/config.go [hunk 11]
```diff
@@ -22,11 +22,16 @@ import (
 // WorkflowControllerConfig contain the configuration settings for the workflow controller
 type WorkflowControllerConfig struct {
 	// ExecutorImage is the image name of the executor to use when running pods
+	// DEPRECATED: use --executor-image flag to workflow-controller instead
```
Comment (maxkramer) [inline workflow/controller/config.go]: Do you mean --executor instead of --executor-image?

File: workflow/artifacts/s3/s3.go [hunk 12]
```diff
@@ -72,7 +75,7 @@ func (s3Driver *S3ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string
 func (s3Driver *S3ArtifactDriver) Save(path string, outputArtifact *wfv1.Artifact) error {
```
Comment (maxkramer) [inline workflow/artifacts/s3/s3.go]: As discussed on Slack, would it be worth adding an info log to stat the file size for debug purposes, in the event that we see a discrepancy between the file on disk and the file in the S3 bucket?

File: workflow/executor/pns/pns.go (hunk 4 above)
Comment (jessesuen) [reply workflow/executor/pns/pns.go]: Which version kubernetes? Process namespace sharing is beta on k8s 1.12. On lower k8s versions, it requires a feature gate.

Comment (dtaniwaki) [reply workflow/executor/pns/pns.go]: I use Kubernetes v1.12.3, but it should be `Process Namespace Sharing is a beta feature that is enabled by default.` based on [the document](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/).

I can see other cotainer's processes by `ps aux` but cannot access the filesystem unless the process is owned by the user.

**Comment by dtaniwaki** [comment]
Can we extract code which is unrelated to PNS from this PR so we can avoid many conflicts with other PR and accelerate the speed of merging other good PRs?
For example, the change for the command line and the code for the output artifacts volumes can be extracted respectively. What do you think?

File: workflow/executor/pns/pns.go (hunk 8 above)
Comment (jessesuen) [reply workflow/executor/pns/pns.go]: this was changed to best effort/warn

File: workflow/controller/config.go (hunk 11 above)
Comment (jessesuen) [reply workflow/controller/config.go]: The flag is actually `--executor-image` in the CLI:

```golang
	command.Flags().StringVar(&executorImage, "executor-image", "", "Executor image to use (overrides value in configmap)")
```

File: workflow/controller/workflowpod.go [hunk 13]
```diff
@@ -138,6 +123,11 @@ func (woc *wfOperationCtx) createWorkflowPod(nodeName string, mainCtr apiv1.Cont
 		}
 		pod.Spec.Containers = append(pod.Spec.Containers, *waitCtr)
 	}
+	// NOTE: the order of the container list is significant. kubelet will pull, create, and start
```
[the remark this answers is not in the document]
Comment (jessesuen) [reply workflow/controller/workflowpod.go]: Actually you can verify the behavior by looking at kubernetes events. It iterates the containers in the order they appear:
```
LAST SEEN   TYPE      REASON               OBJECT                            MESSAGE
9s          Normal    Scheduled            pod/hello-world-ls27f             Successfully assigned default/hello-world-ls27f to minikube
7s          Normal    Pulling              pod/hello-world-ls27f             Pulling image "jessesuen/argoexec:latest"
3s          Normal    Pulled               pod/hello-world-ls27f             Successfully pulled image "jessesuen/argoexec:latest"
3s          Normal    Created              pod/hello-world-ls27f             Created container wait
3s          Normal    Started              pod/hello-world-ls27f             Started container wait
3s          Normal    Pulling              pod/hello-world-ls27f             Pulling image "docker/whalesay:latest"
1s          Normal    Pulled               pod/hello-world-ls27f             Successfully pulled image "docker/whalesay:latest"
1s          Normal    Created              pod/hello-world-ls27f             Created container main
1s          Normal    Started              pod/hello-world-ls27f             Started container main
```

File: workflow/executor/pns/pns.go (hunk 4 above)
Comment (jessesuen) [reply workflow/executor/pns/pns.go]: Good catch! If the main container is not the same user, then you are correct that this won't work. The fix is that we need to grant the wait sidecar SYS_PTRACE capabilities.
```
    securityContext:
      capabilities:
        add:
        - SYS_PTRACE
```

Comment (jessesuen) [reply workflow/executor/pns/pns.go]: OK I fixed the SYS_PTRACE issue, and also a timing related issue. PNS should be working much more reliably now.

Comment (dtaniwaki) [reply workflow/executor/pns/pns.go]: Cool. Let me try it. 👍

File: workflow/controller/workflowpod.go [hunk 14]
```diff
@@ -622,6 +628,41 @@ func (woc *wfOperationCtx) addInputArtifactsVolumes(pod *apiv1.Pod, tmpl *wfv1.T
...
+	if mainCtrIndex == -1 || waitCtrIndex == -1 {
+		panic("Could not find main or wait container in pod spec")
+	}
+	mainCtr = &pod.Spec.Containers[mainCtrIndex]
+	waitCtr := &pod.Spec.Containers[waitCtrIndex]
+
+	for _, mnt := range mainCtr.VolumeMounts {
+		mnt.MountPath = filepath.Join(common.ExecutorMainFilesystemDir, mnt.MountPath)
+		// ReadOnly is needed to be false for overlapping volume mounts
+		mnt.ReadOnly = false
```
Comment (louisblin) [inline workflow/controller/workflowpod.go:660]: This clashes with PSPs using read-only `allowedHostPaths`.

If a workflow container has overlapping volume mounts, will it not have to disable readOnly itself and won't `mnt.ReadOnly` already be set to the right value?
